daemon/pool.go

```diff
@@ -73,123 +100,445 @@ func NewDriverPool(factory FactoryFunction) *DriverPool {
-		if err != nil {
-			return err
+// runSpawn is a goroutine responsible for spawning new instances in the background.
+func (dp *DriverPool) runSpawn() {
+	defer dp.wg.Done()
+	for {
+		select {
+		case <-dp.stop:
+			return
+		case <-dp.spawn:
+			dp.spawnOne()
 		}
+	}
+}
 
-		dp.drivers.Lock()
-		dp.drivers.byID[d.ID()] = d
-		dp.drivers.Unlock()
-		dp.queue.Put(d)
-		dp.stats.instances.Add(1)
+// peekIdle tries to get an idle driver from the pool. It won't wait for the driver to
+// become idle, instead it will return false if there are no idle drivers.
+func (dp *DriverPool) peekIdle() (Driver, bool) {
+	dp.drivers.RLock()
+	n := len(dp.drivers.idle)
+	dp.drivers.RUnlock()
+	if n == 0 {
+		return nil, false
 	}
+	dp.drivers.Lock()
+	defer dp.drivers.Unlock()
+	for d := range dp.drivers.idle {
+		delete(dp.drivers.idle, d)
+		return d, true
+	}
+	return nil, false
+}
 
-	return nil
+// setIdle returns the driver to an idle state.
+func (dp *DriverPool) setIdle(d Driver) {
+	dp.drivers.Lock()
+	defer dp.drivers.Unlock()
+	dp.drivers.idle[d] = struct{}{}
 }
 
-func (dp *DriverPool) del(n int) error {
-	for i := 0; i < n; i++ {
-		d, more := dp.queue.Get()
-		if !more {
-			return ErrPoolClosed.New()
+// killDriver stops are removes the driver from the queue.
+func (dp *DriverPool) killDriver(d Driver) {
+	dp.drivers.Lock()
+	delete(dp.drivers.all, d)
+	delete(dp.drivers.idle, d)
+	dp.running.Add(-1)
+	dp.exited.Add(1)
+	dp.drivers.Unlock()
+
+	if err := d.Stop(); err != nil {
+		dp.Logger.Errorf("error removing stopped driver: %s", err)
+	}
+}
+
+// scaleDiff returns current difference between the target number of instances and the
+// current number of running instances.
+func (dp *DriverPool) scaleDiff() int {
+	total := dp.running.Value()
+	return dp.target.Value() - total
+}
+
+// scale the driver pool to the current target number of instances.
+func (dp *DriverPool) scale() {
+	dn := dp.scaleDiff()
+	if dn == 0 {
+		return
+	}
+
+	// rescaleLater interrupts the scaling and serves the request first.
+	// It will make sure to continue scaling later.
+	rescaleLater := func(req driverRequest) {
+		select {
+		case dp.rescale <- struct{}{}:
+		default:
+		}
+		dp.waitOrScale(req)
+	}
+	if dn < 0 {
+		// scale down
+		for i := 0; i < -dn; i++ {
+			if d, ok := dp.peekIdle(); ok {
+				dp.killDriver(d)
+				continue
+			}
+			select {
+			case <-dp.stop:
+				return
+			case req := <-dp.get:
+				rescaleLater(req)
+				return
+			case d := <-dp.put:
+				dp.killDriver(d)
+			}
 		}
+		return
+	}
+	// scale up
+	for i := 0; i < dn; i++ {
+		select {
+		case req := <-dp.get:
+			rescaleLater(req)
+			return
+		case <-dp.stop:
+			return
+		case d := <-dp.put:
+			// received some existing instance
+			dp.setIdle(d)
+			i--
+			dn = dp.scaleDiff()
+		case dp.spawn <- struct{}{}:
+			// spawn request sent to the spawn goroutine
+			select {
+			case <-dp.stop:
+				return
+			case d := <-dp.put:
+				dp.setIdle(d)
+			case <-dp.spawnErr:
```

Comment:
If it's already been logged and we aren't going to do anything about it, can we not bother sending it at all?

Reply:
It is unused only in this code path. The reason for this channel is that an error may or may not persist. If it does, we are not worried about receiving a specific one and may drop some of them when it's received by a background goroutine. But at the same time, we can return it to the client if he asks for a new driver. And if an error does not persist, it will only be logged and the client won't notice any disruption.